### PR TITLE
Rewrite the ssh repository

### PR DESCRIPTION
**Pull request by popcornylu**
Use ssh library rather then ssh command

### CODE REVIEW

File: docs/content/en/backends/ssh.md [hunk 1]
```diff
@@ -5,52 +5,113 @@ weight: 2
...
+    host bastion
+        hostname mybastion.hosts
+        user myname
+        identityfile ~/id_ed25519
+
+    host myserver
+        hostname myserver.hosts
+        user myname
+        identityfile ~/id_ed25519
+        ProxyCommand ssh bastion -W %h:%p
```
Comment (wcchang1115) [inline docs/content/en/backends/ssh.md:69]: %p need port setting?

Comment (popcornylu) [reply docs/content/en/backends/ssh.md:69]: The two are the same, it just a command line to run

File: docs/content/en/backends/ssh.md [hunk 2]
```diff
@@ -5,52 +5,113 @@ weight: 2
...
+
+
+| Name | Description | Default value |
+| --- | --- | --- |
+| `SSH_USER` | The login user | The current user. |
+| `SSH_PASSWORD` | The passowrd to be used for password authentication |  |
+| `SSH_PORT` | The port of the ssh server | 22 |
+| `SSH_IDENTITY_FILE` | The identity file to be used for pubkey authentication |  |
+| `SSH_KEY_PASSPHRASE` | The passphrase of the identity key  |  |
+| `SSH_SRTICT_HOST_KEY_CHECKING` | Set the value to `no` to disable the key checking | |
```
Comment (wcchang1115) [inline docs/content/en/backends/ssh.md:108]: a typo: `SSH_STRICT_HOST_KEY_CHECKING`

File: internal/repository/ssh.go [hunk 3]
```diff
@@ -1,257 +1,543 @@
...
+		}
 	}
 
-	script = os.Expand(script, func(k string) string {
-		if value, ok := expandMap[k]; ok {
-			return value
-		} else {
-			return "$" + k
+	// host key callbacks: knownhosts
+	if value := os.Getenv("SSH_SRTICT_HOST_KEY_CHECKING"); value != "" {
```
Comment (wcchang1115) [inline internal/repository/ssh.go:138]: a typo: `SSH_STRICT_HOST_KEY_CHECKING`

File: internal/repository/ssh.go [hunk 4]
```diff
@@ -1,257 +1,543 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse SSH_PORT: %s", err.Error())
+		}
 	}
 
-	taskLocal := func(ctx context.Context) error {
-		_, err := CopyWithMeter(dest, src, m)
+	// ssh agent
+	var agentClient agent.ExtendedAgent
+	if agentSock := os.Getenv("SSH_AUTH_SOCK"); agentSock != "" {
```
Comment (wcchang1115) [inline internal/repository/ssh.go:166]: `SSH_AUTH_SOCK` explaination needed

File: internal/repository/ssh.go [hunk 5]
```diff
@@ -1,257 +1,543 @@
...
 	if err != nil {
 		return err
 	}
 
+	knownHostFile := filepath.Join(currentUser.HomeDir, ".ssh", "known_hosts")
+	log.Debugln("check known hosts by file " + knownHostFile)
+
+	knownhostCallback, err := knownhosts.New(knownHostFile)
+	if err != nil {
+		log.Debug("cannot load knownhost file: " + err.Error())
```
Comment (wcchang1115) [inline internal/repository/ssh.go:298]: should it return `err` here?

File: docs/content/en/backends/ssh.md (hunk 1 above)
Comment (wcchang1115) [reply docs/content/en/backends/ssh.md:69]: 1. yeah, `%p` will be 22 if no port is specified.
2. both work!

File: internal/repository/ssh.go (hunk 5 above)
Comment (popcornylu) [reply internal/repository/ssh.go:298]: No. It is just warning. known_hosts file could be absent.

**Pull request merged**